icu4c/source/common/umutex.h

```diff
@@ -54,15 +54,19 @@ U_NAMESPACE_END
 
 #include <atomic>
 
-U_NAMESPACE_BEGIN
-
 // Export an explicit template instantiation of std::atomic<int32_t>. 
 // When building DLLs for Windows this is required as it is used as a data member of the exported SharedObject class.
 // See digitlst.h, pluralaffix.h, datefmt.h, and others for similar examples.
 #if U_PF_WINDOWS <= U_PLATFORM && U_PLATFORM <= U_PF_CYGWIN
+  #if defined(__clang__) && __has_warning("-Winstantiation-after-specialization")
+    // Suppress the warning that the explicit instantiation after explicit specialization has no effect.
+    #pragma clang diagnostic ignored "-Winstantiation-after-specialization"
+  #endif
 template struct U_COMMON_API std::atomic<int32_t>;
```

Comment:
We are  seeing a related compile problem in a Windows environment (with the Microsoft 2015 compiler) that is fixed by
`template <> struct ...`
(adding the empty angle brackets.) As best I understand templates, this means the same thing, but I'm not an expert. If it seems ok, would you mind adding this too?

Reply:
@aheninger What was the error that you were seeing with MSVC/VS2015 ?
I tested building with VS2015, and I don't see any issues with using the syntax: `template struct`.
Thanks!

Reply:
Looking a little closer, the problem was with clang, building for a Windows environment, using headers from VS 2015.
"umutex.h:64:35: error: explicit instantiation of 'atomic' not in a namespace enclosing 'std'."  And similar for other explicit instantiations. Adding the <> got rid of this error, but others followed. I ended up omitting the explicit instantiations altogether by modifying the platform conditional around  them to check for clang.

Reply:
Thanks Andy!
Did moving the `U_NAMESPACE_BEGIN` to below the std::atomic line have any effect? (I'm curious if that might have resolved the error).
Though perhaps just removing the line altogether if the explicit instantiations wasn't needed might be better (i.e.: perhaps you were building a single combined DLL instead of having two DLLs).

Reply:
Moving the U_NAMESPACE_BEGIN to below the std::atomic line gives a different warninng/error:

umutex.h:63:35: error: explicit instantiation of 'atomic<int>' that occurs after an explicit specialization has no effect [-Werror,-Winstantiation-after-specialization]
template struct U_COMMON_API std::atomic<int32_t>;

I'm not sure exactly how this build is being packaged, multiple DLLs or not.

Reply:
Thanks Andy!
Note: The change that is in this current pull-request contains a suppression of that exact warning/error. (Since the claim is that it would have no effect, I think suppressing the warning might be fine.)